native/wasmex/src/instance.rs

```diff
@@ -216,128 +215,141 @@ pub fn call_exported_function(
+                .params()
+                .collect::<Vec<ValType>>(),
+            given_params,
+        );
+        let function_params = match function_params_result {
+            Ok(vec) => map_wasm_values_to_vals(&vec),
+            Err(reason) => return env.error_tuple(&reason).encode(env),
+        };
+        let results_count = function.ty(&*store_or_caller).results().len();
+        let mut results = vec![Val::null_extern_ref(); results_count];
+        let call_result = function.call(
+            &mut *store_or_caller,
+            function_params.as_slice(),
+            &mut results,
+        );
+        match call_result {
+            Ok(_) => (),
+            Err(err) => {
+                let reason = format!("{err}");
+                if let Ok(trap) = err.downcast::<Trap>() {
+                    return env
+                        .error_tuple(format!(
+                            "Error during function excecution ({trap}): {reason}"
+                        ))
+                        .encode(env);
+                } else {
+                    return env
+                        .error_tuple(format!("Error during function excecution: {reason}"))
+                        .encode(env);
+                }
             }
-        })
-    }
-    make_tuple(
-        thread_env,
-        &[
-            atoms::returned_function_call().encode(thread_env),
-            make_tuple(
-                thread_env,
-                &[
-                    atoms::ok().encode(thread_env),
-                    return_values.encode(thread_env),
-                ],
-            ),
-            from,
-        ],
-    )
+        };
+        let mut return_values: Vec<Term> = Vec::with_capacity(results_count);
+        for value in results.iter().cloned() {
+            return_values.push(match value {
+                Val::I32(i) => i.encode(env),
+                Val::I64(i) => i.encode(env),
+                Val::F32(i) => f32::from_bits(i).encode(env),
+                Val::F64(i) => f64::from_bits(i).encode(env),
+                Val::V128(i) => rustler::BigInt::from(i.as_u128()).encode(env),
+                Val::FuncRef(_) => {
+                    return env
+                        .error_tuple("unable_to_return_func_ref_type")
+                        .encode(env)
+                }
+                Val::ExternRef(_) => {
+                    return env
+                        .error_tuple("unable_to_return_extern_ref_type")
+                        .encode(env)
+                }
+                Val::AnyRef(_) => {
+                    return env.error_tuple("unable_to_return_any_ref_type").encode(env)
+                }
+            })
+        }
+
+        // Return nil for void functions (no return values), otherwise return the list
+        let result_term = if return_values.is_empty() {
+            atoms::__nil__().encode(env)
+        } else {
+            return_values.encode(env)
+        };
```

Comment:
I thought about returning just `nil` if results are empty and I very much see the convenience. However, at the Rust-level I like to be as close to wasmtime as possible and warm/wasmtime thinks in a _list_ of return values, not just one.

replacing an empty results list with `nil` is a leaky abstraction IMO, so I'll change this back.